### PR TITLE
Use the transaction builder to implement z_sendmany.

### LINKED ISSUE #5433: Replace manual construction of Sprout-inclusive transactions in asyncrpcoperation_sendmany.cpp with transaction builder.
At present, asyncrpcoperation_sendmany is quite complex due to the way it handles different subsets of transaction construction. The transaction builder now supports Sprout transactions, but asyncrpcoperation_sendmany was never updated to take advantage of this fact. Simplifying the internals of asyncrpcoperation_sendmany to use the transaction builder for all cases will dramatically simplify and make safer the process of adding unified address support.

### PR DESCRIPTION
**Pull request by nuttycom**
Fixes #5433

### CODE REVIEW

File: qa/rpc-tests/test_framework/util.py [hunk 1]
```diff
@@ -563,7 +563,7 @@ def assert_equal(expected, actual, message=""):
     if expected != actual:
         if message:
             message = "; %s" % message 
-        raise AssertionError("(left == right)%s\n  left: <%s>\n right: <%s>" % (message, str(expected), str(actual)))
+        raise AssertionError("(expected == actual)%s\n expected: <%s>\n   actual: <%s>" % (message, str(expected), str(actual)))
```
Comment (daira) [inline qa/rpc-tests/test_framework/util.py:566]: Even though the arguments are called `expected` and `actual`, it's not guaranteed that callers will actually use them that way. If a caller swaps around their meanings, this assertion message could become very confusing. I would revert this change.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 2]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
-                }
-            }
+            throw JSONRPCError(
+                    RPC_WALLET_INSUFFICIENT_FUNDS,
+                    strprintf(
+                        "Insufficient funds: have %s, need %s",
+                        FormatMoney(spendable.Total()), FormatMoney(txValues.targetAmount))
+                        + (allowTransparentCoinbase ? "" :
+                            "; if you are attempting to shield transparent coinbase funds, "
+                            "ensure that you have specified only a single recipient address.")
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:282]: Change the message to reflect what is enforced now.

File: qa/rpc-tests/wallet_shieldingcoinbase.py [hunk 3]
```diff
@@ -86,17 +86,19 @@ def run_test (self):
             myopid = self.nodes[3].z_sendmany(mytaddr, recipients)
         except JSONRPCException as e:
             errorString = e.error['message']
-        assert_equal("Invalid from address: does not belong to this node, spending key not found.", errorString);
+        assert_equal("Invalid from address, no spending key found for address", errorString);
 
         # This send will fail because our wallet does not allow any change when shielding a coinbase utxo,
         # as it's currently not possible to specify a change address in z_sendmany.
```
Comment (daira) [inline qa/rpc-tests/wallet_shieldingcoinbase.py:92]: It's not that the wallet doesn't allow change; it's a consequence of this consensus rule:
> A transaction with one or more transparent inputs from coinbase transactions MUST have no transparent
outputs (i.e. `tx_out_count` MUST be 0). Inputs from coinbase transactions include Founders’ Reward outputs
and funding stream outputs.

Since the input is transparent, the change also can't be transparent, because that would be a transparent output. If we were to specify a shielded address for "change", that would contradict the definition of "change" as funds that are sent to an address that is also spent from in the same transaction.

(Technically, there could be another shielded input that is also where change goes to, but that's not why *this* send will fail, even if the wallet were to support that, because it has no such shielded input.)

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 4]
```diff
@@ -36,104 +36,64 @@
...
-
-    KeyIO keyIO(Params());
-
-    useanyutxo_ = fromAddress == "ANY_TADDR";
-    if (useanyutxo_) {
-        isfromtaddr_ = true;
-    } else {
-        auto address = keyIO.DecodePaymentAddress(fromAddress);
-        if (address.has_value()) {
+    assert(mindepth_ >= 0);
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:59]: Why not change `minDepth` and `mindepth_` to unsigned?

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 5]
```diff
@@ -36,104 +36,64 @@
...
-    for (size_t i = 0; i < outputMap.size(); i++) {
-        if (outputMap[i].get_int() == n) {
-            return i;
-        }
-    }
-
-    throw std::logic_error("n is not present in outputmap");
-}
-
 AsyncRPCOperation_sendmany::AsyncRPCOperation_sendmany(
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:46]: Document the preconditions on arguments.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 6]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
+            return true;
+        },
+        [&](const PaymentAddress& addr) {
+            return false;
+        }
+    }, paymentSource);
+}
+
+// Construct and send the transaction, returning the resulting txid.
+// Errors in transaction construction will throw.
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:192]:
```suggestion
// Construct and send the transaction, returning the resulting txid if successful.
// Errors in transaction construction will either throw or return `std::nullopt`.
```
(if this is correct; I'm not sure).

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 7]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
+//
 // Notes:
 // 1. #1159 Currently there is no limit set on the number of joinsplits, so size of tx could be invalid.
 // 2. #1360 Note selection is not optimal
 // 3. #1277 Spendable notes are not locked, so an operation running in parallel could also try to use them
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:197]:
```suggestion
// 1. #1159 Currently there is no limit set on the number of shielded elements (JoinSplits, Spends, or Outputs),
//    which could make the tx too large.
// 2. #1360 Note selection is not optimal.
// 3. #1277 Spendable notes are not locked, so an operation running in parallel could also try to use them.
// 4. #1614 Anchors are chosen at the most recent block; this is unreliable and leaks information in case of rollback.
// 5. #3615 There is no padding of inputs or outputs, which may leak information. This differs from the Rust tx builder.
```

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 8]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
+                // unreachable; currently disallowed by checks at construction
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Sending to Sprout is disabled.");
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                txValues.z_outputs_total += recipient.amount;
+                shieldedRecipients += 1;
+                if (isfromsprout_ && !allowRevealedAmounts_) {
+                    throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
+                            "Sending from Sprout to Sapling is not enabled by default because it will "
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:228]:
```suggestion
                            "Sending between shielded pools is not enabled by default because it will "
```
We should make the error message reflect what the generalized policy will be, since we already know it.

File: qa/rpc-tests/test_framework/util.py (hunk 1 above)
Comment (nuttycom) [reply qa/rpc-tests/test_framework/util.py:566]: Meh, okay. I made this change specifically because I found `left` and `right` confusing in the resulting error messages. I would rather standardize on the order of arguments, but I guess that doing so requires checking every call site, and I'm not going to make those changes here.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 9]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
+    // Find spendable inputs, and select a minimal set of them that
+    // can supply the required target amount.
+    auto spendable = FindSpendableInputs(allowTransparentCoinbase);
+    if (!spendable.LimitToAmount(txValues.targetAmount, dustThreshold)) {
+        CAmount changeAmount{spendable.Total() - txValues.targetAmount};
+        if (changeAmount > 0 && changeAmount < dustThreshold) {
+            // TODO: this condition is silly; we should provide the option for the caller to
+            // forego change (definitionally amount below the dust amount) and send the
+            // extra to the recipient or the miner fee to avoid creating dust change, rather
+            // than prohibit them from sending entirely in this circumstance.
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:264]: For dust shielded notes, that potentially leaks information about the note value, so we should do this carefully.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 10]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
-        zOutputsDeque.push_back(o);
-    }
 
+    // Find Sprout witnesses
     // When spending notes, take a snapshot of note witnesses and anchors as the treestate will
     // change upon arrival of new blocks which contain joinsplit transactions.  This is likely
     // to happen as creating a chained joinsplit transaction can take longer than the block interval.
-    if (z_sprout_inputs_.size() > 0) {
+    // So, we need to take locks on cs_main and pwalletMain->cs_wallet so that the witnesses aren't
+    // updated.
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:493]: The fact that we lock and unlock `cs_main` and `cs_wallet` multiple times, means that the Sprout and Sapling anchors can come from different blocks. This is technically a wallet implementation distinguisher. I don't think it's a big deal in practice, but we could do these in a single `lock/unlock`.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 9 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:264]: Suppose that you're in this situation: you have a 100 ZEC note, and your counterparty requests a 99.99998999 ZEC payment. It's a little odd that that z_sendmany will not currently permit you to pay this bill unless you explicitly send 99.99999000. One alternative is to include this suggestion in the error message; then the user can explicitly opt-in to it by recreating their transaction, but that's just a hassle for the user when we could provide a flag to allow them to ask the machine to do it.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 11]
```diff
@@ -191,791 +149,508 @@ void AsyncRPCOperation_sendmany::main() {
...
-                jsAnchor = inputAnchor;
-            }
+    // Build the transaction
+    auto buildResult = builder_.Build();
+    auto tx = buildResult.GetTxOrThrow();
 
-            // Add spendable notes as inputs
-            std::copy(vInputNotes.begin(), vInputNotes.end(), std::back_inserter(info.notes));
-        }
+    UniValue sendResult = SendTransaction(tx, keyChange, testmode);
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:525]: I would split out a part that creates and returns the transaction without sending it, so that it can more easily be tested independently. (I know this part still takes global locks and reads from the blockchain.)

Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:525]: I think this is a good idea, but I think the right thing to do here is factor out a wallet-level transaction builder that then gets reused in z_shieldcoinbase, z_mergetoaddress etc. that implements the wallet rules. The current implementation retains the behavior that sending to the network is part of the async operation.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 10 above)
Comment (str4d) [reply src/wallet/asyncrpcoperation_sendmany.cpp:493]: We can fix this in a subsequent refactor, after this PR (and probably after extracting common coin selection logic from the three main RPCs).

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 9 above)
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:264]: Yes, but we should not make fingerprinting attacks easier.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 10 above)
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:493]: Please add a TODO for that.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 12]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+    auto sproutIt = sproutNoteEntries.begin();
+    while (sproutIt != sproutNoteEntries.end() && amountRequired > 0) {
+        amountRequired -= sproutIt->note.value();
+        ++sproutIt;
+        // if the total would result in change below the dust threshold,
+        // adjust the amountRequired so that we select additional inputs
+        if (amountRequired < 0 && amountRequired > -dustThreshold) {
+            amountRequired += dustThreshold;
+            dustThreshold = amountRequired;
+        }
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:593]: Consider the following example:
```
dustThreshold = 10
amountRequired = 1000
note1.value = 1001
note2.value = 10
```
After the first iteration we have `amountRemaining = dustThreshold = 9`. Then on the second iteration we subtract 10 from `amountRemaining` and hit the "change below the dust threshold" condition again and we have `amountRemaining = dustThreshold = 8`, having spent 1011. Spending 1011, with 1000 going to the recipient and 11 to change, is already sufficient because 11 is not dust. But the algorithm does not recognize that and wants to spend at least 8 more.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 9 above)
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:264]: The real solution to dust issues is #2133 :-p

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 13]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
-
-    // If using the TransactionBuilder, we only want Sapling notes.
-    // If not using it, we only want Sprout notes.
-    // TODO: Refactor `GetFilteredNotes()` so we only fetch what we need.
-    if (isUsingBuilder_) {
-        sproutEntries.clear();
-    } else {
-        saplingEntries.clear();
+void SpendableInputs::LogInputs(const AsyncRPCOperationId& id) const {
+    for (const auto& utxo : utxos) {
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:647]: Check that this can't be a performance bottleneck. Some nodes have *lots* of UTXOs. Now that I think about it, this may be why we didn't originally do this logging.

File: src/wallet/test/rpc_wallet_tests.cpp [hunk 14]
```diff
@@ -1498,8 +1372,8 @@ BOOST_AUTO_TEST_CASE(rpc_z_sendmany_taddr_to_sapling)
     auto builder = TransactionBuilder(consensusParams, nextBlockHeight, pwalletMain);
     mtx = CreateNewContextualCMutableTransaction(consensusParams, nextBlockHeight);
 
-    std::vector<SendManyRecipient> recipients = { SendManyRecipient(zaddr1, 1*COIN, "ABCD") };
-    std::shared_ptr<AsyncRPCOperation> operation( new AsyncRPCOperation_sendmany(builder, mtx, taddr1, {}, recipients, 0) );
+    std::vector<SendManyRecipient> recipients = { SendManyRecipient(pa, 1*COIN, "ABCD") };
+    std::shared_ptr<AsyncRPCOperation> operation(new AsyncRPCOperation_sendmany(builder, taddr, recipients, 1));
```
Comment (str4d) [inline src/wallet/test/rpc_wallet_tests.cpp:1376]: Why is `minconf` set to 1 here now instead of 0?

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 15]
```diff
@@ -1250,7 +701,9 @@ std::array<unsigned char, ZC_MEMO_SIZE> AsyncRPCOperation_sendmany::get_memo_fro
     }
 
     if (rawMemo.size() > ZC_MEMO_SIZE) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Memo size of %d is too big, maximum allowed is %d", rawMemo.size(), ZC_MEMO_SIZE));
+        throw JSONRPCError(
+                RPC_INVALID_PARAMETER,
+                strprintf("Memo size of %d is too big, maximum allowed is %d", rawMemo.size(), ZC_MEMO_SIZE));
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:706]:
```suggestion
                strprintf("Memo size of %d bytes is too big, maximum allowed is %d bytes", rawMemo.size(), ZC_MEMO_SIZE));
```

**Review by str4d — CHANGES_REQUESTED** [review]

**Comment by str4d** [comment]
CI is currently failing `btest`.

**Review by daira — APPROVED** [review]
utACK modulo comments.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 16]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
 
-            LogPrint("zrpcunsafe", "%s: generating note for change (amount=%s)\n",
-                    getId(),
-                    FormatMoney(jsChange)
-                    );
+bool SpendableInputs::LimitToAmount(CAmount amountRequired, CAmount dustThreshold) {
+    // select Sprout notes for spending first
+    std::sort(sproutNoteEntries.begin(), sproutNoteEntries.end(),
+        [](SproutNoteEntry i, SproutNoteEntry j) -> bool {
+            return i.note.value() > j.note.value();
```
[the remark this answers is not in the document]
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:582]: We should make a separate deliberate decision about this, not change it as part of this refactoring. We have a [whole bunch of tickets on this subject](https://github.com/zcash/zcash/labels/note%20selection%20and%20shielded%20tx%20construction); in particular see #4332 and #1961.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 12 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:593]: Ah, right, because we don't have any explicit transition to a separate "avoiding dust" mode if I try to reuse `amountRequired` for this purpose.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 17]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
-                throw JSONRPCError(RPC_WALLET_INSUFFICIENT_FUNDS, "Insufficient transparent funds, no UTXOs found for taddr from address.");
-            }
+    txValues.targetAmount = sendAmount + fee_;
+
+    builder_.SetFee(fee_);
+
+    // Only select coinbase if we are spending from at most a single t-address.
+    bool allowTransparentCoinbase =
+        !IsFromAnyTaddr(paymentSource_) && // allow coinbase inputs from at most a single t-addr
+        transparentRecipients == 0; // cannot send transparent coinbase to transparent recipients
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:249]: We don't allow change when spending coinbase funds anyway (even before these changes), so we just check that all the target amount is equal to the total of the selected UTXOs.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 18]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+
+    // Set the dust threshold so that we can select enough inputs to avoid
+    // creating dust change amounts.
+    CAmount dustThreshold{DefaultDustThreshold()};
+
+    // Find spendable inputs, and select a minimal set of them that
+    // can supply the required target amount.
+    auto spendable = FindSpendableInputs(allowTransparentCoinbase);
+    if (!spendable.LimitToAmount(txValues.targetAmount, dustThreshold)) {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:258]: I wanted something that reflected the fact that this is mutating the object it's being called on. I would have called it `SelectAmount` if it returned a new immutable value containing just the subset of inputs. So I don't like that name, but am open to something else that has the "mutably limit" meaning.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 19]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+    // associated OVK
+    std::visit(match {
+        [&](const FromAnyTaddr& fromAny) {
+            ovk = getDefaultOVK();
+            setTransparentChangeRecipient();
+        },
+        [&](const PaymentAddress& addr) {
+            std::visit(match {
+                [&](const libzcash::SproutPaymentAddress& addr) {
+                    ovk = getDefaultOVK();
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:390]: Once we have UAs we just use the external OVK for the transparent address for the appropriate account, or the legacy account otherwise. z_sendmany was previously using the transparentOvkForShielding for sprout as well, I believe, so I just maintained that here.

File: qa/rpc-tests/wallet_shieldingcoinbase.py (hunk 3 above)
Comment (nuttycom) [reply qa/rpc-tests/wallet_shieldingcoinbase.py:92]: This wasn't my comment, obviously, but I've fixed it and added a note that this test is expected to start failing after UA integration (which will send the change to the most recent shielded pool corresponding to the account corresponding to the "from" address.)

File: src/wallet/rpcwallet.cpp [hunk 20]
```diff
@@ -3680,14 +3679,90 @@ UniValue z_getoperationstatus_IMPL(const UniValue& params, bool fRemoveFinishedO
...
 #define CTXOUT_REGULAR_SIZE     34
 
+size_t EstimateTxSize(
+        const PaymentSource& paymentSource,
+        const std::vector<SendManyRecipient>& recipients) {
+    int nextBlockHeight = chainActive.Height() + 1;
+    CMutableTransaction mtx;
+    mtx.fOverwintered = true;
+    mtx.nVersionGroupId = SAPLING_VERSION_GROUP_ID;
+    mtx.nVersion = SAPLING_TX_VERSION;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:3689]: Opened https://github.com/zcash/zcash/issues/5465

File: src/wallet/rpcwallet.cpp [hunk 21]
```diff
@@ -3720,15 +3795,16 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
...
     // Check that the from address is valid.
     auto fromaddress = params[0].get_str();
+    PaymentSource paymentSource;
     bool fromTaddr = false;
-    bool fromSapling = false;
     bool fromSprout = false;
-    KeyIO keyIO(Params());
+    bool fromSapling = false;
     if (fromaddress == "ANY_TADDR") {
-        fromTaddr = true;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:3731]: There's the rub; I forgot to remove `fromTaddr`!

File: src/wallet/rpcwallet.cpp [hunk 22]
```diff
@@ -3993,33 +3959,14 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
     o.pushKV("fee", std::stod(FormatMoney(nFee)));
     UniValue contextInfo = o;
 
-    if (!fromTaddr || !zaddrRecipients.empty()) {
-        // We have shielded inputs or outputs, and therefore cannot create
-        // transactions before Sapling activates.
-        if (!Params().GetConsensus().NetworkUpgradeActive(nextBlockHeight, Consensus::UPGRADE_SAPLING)) {
-            throw JSONRPCError(
-                RPC_INVALID_PARAMETER, "Cannot create shielded transactions before Sapling has activated");
-        }
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:4002]: There's already a `ThrowIfInitialBlockDownload` at the start of z_sendmany, is that not sufficient?

File: src/wallet/rpcwallet.cpp [hunk 23]
```diff
@@ -3764,190 +3833,80 @@ UniValue z_sendmany(const UniValue& params, bool fHelp)
...
     }
 
     UniValue outputs = params[1].get_array();
-    if (outputs.size()==0)
+    if (outputs.size() == 0) {
         throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, amounts array is empty.");
+    }
 
-    // Keep track of addresses to spot duplicates
-    set<std::string> setAddress;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/rpcwallet.cpp:3774]: I'll restore it.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 4 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:59]: I think the reason that this was a signed value is that it's coming from UniValue::getInt, and I didn't change it.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 24]
```diff
@@ -36,104 +36,64 @@
...
-            }, address.value());
-        } else {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid from address");
+            }, addr);
         }
-    }
+    }, paymentSource);
 
-    if (isfromzaddr_ && minDepth==0) {
+    if ((isfromsprout_ || isfromsapling_) && minDepth==0) {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:96]: Yes, `isfromsprout_` is used later. But my intent is to actually get rid of all the boolean flags in the process of adding UA support; this was just the minimal change that was necessary.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 25]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+public:
     CAmount t_inputs_total{0};
     CAmount z_inputs_total{0};
     CAmount t_outputs_total{0};
     CAmount z_outputs_total{0};
     CAmount targetAmount{0};
-    bool selectedUTXOCoinbase{false};
+
+    CAmount GetChangeAmount() {
+        return t_inputs_total + z_inputs_total - targetAmount;
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:176]: Actually, this is now unused and can just be deleted.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 26]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+            [&](const libzcash::SproutPaymentAddress& addr) {
+                // unreachable; currently disallowed by checks at construction
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Sending to Sprout is disabled.");
+            },
+            [&](const libzcash::SaplingPaymentAddress& addr) {
+                txValues.z_outputs_total += recipient.amount;
+                shieldedRecipients += 1;
+                if (isfromsprout_ && !allowRevealedAmounts_) {
+                    throw JSONRPCError(
+                            RPC_INVALID_PARAMETER,
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:227]: The reasons that I put these checks here is that this is the first point at which we inspect the individual recipients, and we accumulate a bunch of local state in the process. I guess that I can move all that state to be member variables on the class, that's a little ugly but I guess it's fine.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 27]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
-        // Add Sapling spends
-        for (size_t i = 0; i < notes.size(); i++) {
-            if (!witnesses[i]) {
-                throw JSONRPCError(RPC_WALLET_ERROR, "Missing witness for Sapling note");
-            }
-            builder_.AddSaplingSpend(expsk, notes[i], anchor, witnesses[i].value());
+                    ovk = saplingKey.expsk.full_viewing_key().ovk;
+                    builder_.SendChangeTo(addr, ovk);
+                },
+                [&](const auto& other) {
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:400]: I won't, as I've already been implementing this change. :)

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 28]
```diff
@@ -191,791 +149,514 @@ void AsyncRPCOperation_sendmany::main() {
...
+        saplingOutPoints.push_back(t.op);
+        saplingNotes.push_back(t.note);
 
-        // Add transparent outputs
-        for (auto r : t_outputs_) {
-            auto outputAddress = r.address;
-            auto amount = r.amount;
+        libzcash::SaplingExtendedSpendingKey saplingKey;
+        assert(pwalletMain->GetSaplingExtendedSpendingKey(t.address, saplingKey));
+        saplingKeys.push_back(saplingKey);
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:422]: That's true. I think that we should keep this as-is for now, because I want to keep note selection decoupled from transaction construction.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 12 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:598]: Fixed.

File: src/wallet/rpcwallet.cpp (hunk 22 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/rpcwallet.cpp:4002]: Ah yes, that is indeed sufficient.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 18 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/asyncrpcoperation_sendmany.cpp:261]: Maybe `DiscardAboveAmount` then?

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (str4d) [reply src/wallet/asyncrpcoperation_sendmany.cpp:394]: Previously we _couldn't_ send Sprout->Sapling via `z_sendmany`, so there wasn't any OVK used for Sprout (since Sprout itself has no OVK usage, and neither does transparent).

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 29]
```diff
@@ -143,9 +103,6 @@ AsyncRPCOperation_sendmany::AsyncRPCOperation_sendmany(
     } else {
         LogPrint("zrpc", "%s: z_sendmany initialized\n", getId());
     }
-
-    // Enable payment disclosure if requested
-    paymentDisclosureMode = fExperimentalPaymentDisclosure;
```
[the remark this answers is not in the document]
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:148]: Sprout payment disclosures were always an experimental feature, so I'm not sure we should spend much effort on removing them gracefully. To me the point of marking a feature as experimental is to indicate that we won't necessarily do that; we are fully allowed to just yank it.

File: src/wallet/asyncrpcoperation_sendmany.cpp [hunk 30]
```diff
@@ -191,791 +149,525 @@ void AsyncRPCOperation_sendmany::main() {
...
+    // can supply the required target amount.
+    auto spendable = FindSpendableInputs(allowTransparentCoinbase);
+    if (!spendable.LimitToAmount(txValues.targetAmount, dustThreshold)) {
+        CAmount changeAmount{spendable.Total() - txValues.targetAmount};
+        if (changeAmount > 0 && changeAmount < dustThreshold) {
+            // TODO: we should provide the option for the caller to explicitly
+            // forego change (definitionally an amount below the dust amount)
+            // and send the extra to the recipient or the miner fee to avoid
+            // creating dust change, rather than prohibit them from sending
+            // entirely in this circumstance.
```
Comment (daira) [inline src/wallet/asyncrpcoperation_sendmany.cpp:268]:
```suggestion
            // entirely in this circumstance.
            // (Daira disagrees since this may leak note values.)
```

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (daira) [reply src/wallet/asyncrpcoperation_sendmany.cpp:394]: Sprout keys don't have OVKs. This seems fine to me.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 30 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:254]: I will add this comment to the branch that integrates UAs, so as to avoid a CI run.

File: src/wallet/asyncrpcoperation_sendmany.cpp (hunk 5 above)
Comment (nuttycom) [reply src/wallet/asyncrpcoperation_sendmany.cpp:46]: I will also add this documentation to the UA integration branch.

File: qa/rpc-tests/wallet_listreceived.py [hunk 31]
```diff
@@ -226,7 +226,7 @@ def run_test_release(self, release, height):
         assert_equal(3, c[release], "Count of unconfirmed notes should be 3(2 in zaddr1 + 1 in zaddr2)")
 
     def run_test(self):
-        self.run_test_release('sprout', 200)
+        #self.run_test_release('sprout', 200)
         self.run_test_release('sapling', 214)
```
[the remark this answers is not in the document]
Comment (nuttycom) [reply qa/rpc-tests/wallet_listreceived.py:230]: Added to #5467

**Review by daira — APPROVED** [review]
utACK de58fc0b168e129a17eae56913425af123a53fdc
